controllers/drclusters.go

```diff
@@ -189,3 +270,97 @@ func drClustersUndeploy(drpolicy *rmn.DRPolicy, mwu *util.MWUtil) error {
 
 	return nil
 }
+
+func drClusterListMustHaveS3Profiles(drpolicies rmn.DRPolicyList,
+	clusterName string,
+	ignorePolicy *rmn.DRPolicy) sets.String {
+	mustHaveS3Profiles := sets.String{}
+
+	for idx, drpolicy := range drpolicies.Items {
+		// Skip the policy being ignored (used for delete)
+		if (ignorePolicy != nil) && (ignorePolicy.ObjectMeta.Name == drpolicy.ObjectMeta.Name) {
+			continue
+		}
+
+		for _, cluster := range util.DrpolicyClusterNames(&drpolicies.Items[idx]) {
+			// Skip if not the current cluster
+			if cluster != clusterName {
+				continue
+			}
+
+			// Add all S3Profiles across clusters in the policy to the current cluster
+			for _, managedCluster := range drpolicy.Spec.DRClusterSet {
+				mustHaveS3Profiles = mustHaveS3Profiles.Insert(managedCluster.S3ProfileName)
+			}
+
+			break
+		}
+	}
+
+	return mustHaveS3Profiles
+}
+
+// drClusterListMustHaveSecrets lists s3 secrets that must exist on the passed in clusterName
+// It optionally ignores a specified ignorePolicy, which is typically useful when a policy is being
+// deleted.
+func drClusterListMustHaveSecrets(
+	drpolicies rmn.DRPolicyList,
+	clusterName string,
+	ignorePolicy *rmn.DRPolicy,
+	ramenConfig *rmn.RamenConfig) sets.String {
+	mustHaveS3Secrets := sets.String{}
+
+	mustHaveS3Profiles := drClusterListMustHaveS3Profiles(drpolicies, clusterName, ignorePolicy)
+
+	// Determine s3Secrets that must continue to exist on the cluster, based on other profiles
+	// that should still be present. This is done as multiple profiles MAY point to the same secret
+	for _, s3Profile := range ramenConfig.S3StoreProfiles {
+		if mustHaveS3Profiles.Has(s3Profile.S3ProfileName) {
+			mustHaveS3Secrets = mustHaveS3Secrets.Insert(s3Profile.S3SecretRef.Name)
+		}
+	}
+
+	return mustHaveS3Secrets
+}
+
+func drClustersUndeploySecrets(
+	drpolicy *rmn.DRPolicy,
+	drpolicies rmn.DRPolicyList,
+	secretsUtil *util.SecretsUtil,
+	ramenConfig *rmn.RamenConfig) error {
+	if !ramenConfig.DrClusterOperator.DeploymentAutomationEnabled ||
+		!ramenConfig.DrClusterOperator.S3SecretDistributionEnabled {
```

Comment:
If a user were to change one or both of these booleans from true to false, after secrets were distributed, I suppose it could leave the policy related resources and secrets around, which is possibly what the user intends.  But I wonder if it should instead cleanup resources that it created (or updated?) on undeploy even if the booleans have been set to false.